phoenix-shared/src/commonMain/kotlin/fr.acinq.phoenix/managers/CurrencyManager.kt

```diff
@@ -40,49 +65,178 @@ class CurrencyManager(
-                    BitcoinPriceRate(
-                        fiatCurrency = fiatCurrency,
-                        price = it.value.last,
-                        source = "blockchain.info",
-                        timestampMillis = Clock.System.now().toEpochMilliseconds(),
-                    )
-                } ?: run {
-                    log.info { "Blockchain.info has new currency: ${it.key}" }
-                    null
+                    if (fiatCurrency == FiatCurrency.USD || fiatCurrency == FiatCurrency.EUR) {
+                        ExchangeRate.BitcoinPriceRate(
+                            fiatCurrency = fiatCurrency,
+                            price = it.value.last,
+                            source = "blockchain.info",
+                            timestampMillis = Clock.System.now().toEpochMilliseconds(),
+                        )
+                    } else {
+                        null
+                    }
                 }
             }
-            .forEach { appDb.saveBitcoinRate(it) }
+            .let { fetchedRates ->
+                appDb.saveExchangeRates(fetchedRates)
+            }
         log.info { "successfully refreshed bitcoin prices from blockchain.info" }
+        true
     } catch (e: Exception) {
         log.error(e) { "failed to refresh bitcoin price rates from blockchain.info ticker: $e" }
+        false
     }
 
-    private suspend fun refreshFromBitso() = try {
-        log.info { "fetching bitcoin prices from bitso.com" }
-        httpClient.get<MxnApiResponse>("https://api.bitso.com/v3/ticker/?book=btc_mxn")
-            .takeIf { it.success }
-            ?.let {
-                BitcoinPriceRate(
-                    fiatCurrency = FiatCurrency.MXN,
-                    price = it.payload.last,
-                    source = "bitso.com",
-                    timestampMillis = Clock.System.now().toEpochMilliseconds(),
+    private suspend fun refreshFromCoinDesk(): Boolean {
+        log.info { "fetching exchange rates from coindesk.com" }
+        // Performance notes:
+        // If we use zero concurrency, by fetching each URL one-at-a-time,
+        // and writing to the database after each one,
+        // then the entire process takes around 46 seconds !
+        // That's an average wait time of 23 seconds before a given fiat currency is ready.
+        // We're attempting to maximize concurrency by fetching all,
+        // and hopefully using http pipelining to speed up the process.
+        // The end result is that the process takes around 10 seconds.
+        val http = HttpClient { engine {
+            threadsCount = 1
+            pipelining = true
+        }}
+        val json = Json {
+            ignoreUnknownKeys = true
+        }
+        val fiatCurrencies = FiatCurrency.values.filter {
+            it != FiatCurrency.USD && it != FiatCurrency.EUR
+        }
+        return fiatCurrencies.map { fiatCurrency ->
```

Comment:
It's like that because I needed the `fiatCurrencies` variable again below:

```kotlin
appDb.saveExchangeRates(fetchedRates)
log.info { "successfully refreshed exchange rates from coindesk.com" }
// We consider the fetch to be successful if we fetched at least 2/3 of the list
fetchedRates.size >= (fiatCurrencies.size * 2 / 3) // ^let
```

However, I'm not particularly happy about this fuzzy logic...

I know it's possible that one currency starts failing, possibly because CoinDesk drops support for it. And I didn't want this possibility to cause excessive refreshes of ALL the currencies. But let me see if I can rework this...

Reply:
Fixed in 29196b0